Daemon crashes on missing .claude.json.lock during session cleanup

## Problem

When the daemon starts a new session and attempts to clean up or reuse a shepherd config directory from a previous run, it crashes with:

```
[ERROR] Daemon error: [Errno 2] No such file or directory: '.loom/claude-config/shepherd-N/.claude.json.lock'
```

This is a fatal error — the daemon process exits entirely rather than handling the missing file gracefully.

## Reproduction

1. Run the daemon, let it spawn shepherds
2. Stop the daemon (graceful or crash)
3. Start the daemon again
4. Daemon attempts to reuse shepherd config directories from the previous run
5. Crashes on the first shepherd whose `.claude.json.lock` was cleaned up between runs

## Expected behavior

The daemon should handle a missing lock file gracefully — either create it, skip the cleanup step, or catch the `FileNotFoundError` and continue.

## Workaround

Pre-create the lock files for all shepherd slots:
```bash
for i in $(seq 1 10); do
    mkdir -p .loom/claude-config/shepherd-$i
    touch .loom/claude-config/shepherd-$i/.claude.json.lock
done
```

## Environment

- Loom 0.4.0
- macOS, Python daemon via loom-tools
- Crash occurs in `agent_spawn.py` or the daemon's session management code

--- a/loom-tools/src/loom_tools/common/claude_config.py
+++ b/loom-tools/src/loom_tools/common/claude_config.py
@@ -424,6 +424,46 @@ def setup_agent_config_dir(agent_name: str, repo_root: Path) -> Path:
     return config_dir
 
 
+def _rmtree_with_retry(path: Path) -> None:
+    """Remove a directory tree, tolerating files vanishing mid-removal.
+
+    Claude Code creates ephemeral lock files (e.g. ``.claude.json.lock``)
+    that may be cleaned up by a dying session between the time
+    ``shutil.rmtree`` lists the directory and when it tries to unlink
+    each entry.  This causes a ``FileNotFoundError`` that crashes the
+    daemon.  See issue #3097.
+
+    Strategy: first attempt uses ``shutil.rmtree`` with an ``onexc``
+    (Python 3.12+) or ``onerror`` callback that ignores
+    ``FileNotFoundError``.  If the directory still exists after that
+    (very unlikely edge case), a second attempt with
+    ``ignore_errors=True`` ensures we never raise.
+    """
+    import sys
+
+    def _onexc(func, fpath, exc):  # noqa: ARG001 — shutil callback signature
+        """Ignore FileNotFoundError; re-raise anything else (onexc API)."""
+        if isinstance(exc, FileNotFoundError):
+            return
+        raise exc
+
+    def _onerror(func, fpath, exc_info):  # noqa: ARG001 — shutil callback signature
+        """Ignore FileNotFoundError; re-raise anything else (onerror API)."""
+        if isinstance(exc_info[1], FileNotFoundError):
+            return
+        raise exc_info[1]
+
+    if sys.version_info >= (3, 12):
+        shutil.rmtree(path, onexc=_onexc)
+    else:
+        shutil.rmtree(path, onerror=_onerror)
+
+    # Belt-and-suspenders: if something else went wrong and the dir
+    # survived, do a final sweep that silences all errors.
+    if path.is_dir():
+        shutil.rmtree(path, ignore_errors=True)
+
+
 def cleanup_agent_config_dir(agent_name: str, repo_root: Path) -> bool:
     """Remove one agent's config directory.
 
@@ -437,7 +477,7 @@ def cleanup_agent_config_dir(agent_name: str, repo_root: Path) -> bool:
     paths = LoomPaths(repo_root)
     config_dir = paths.agent_claude_config_dir(agent_name)
     if config_dir.is_dir():
-        shutil.rmtree(config_dir)
+        _rmtree_with_retry(config_dir)
         return True
     return False
 
@@ -458,6 +498,6 @@ def cleanup_all_agent_config_dirs(repo_root: Path) -> int:
     count = 0
     for child in base_dir.iterdir():
         if child.is_dir():
-            shutil.rmtree(child)
+            _rmtree_with_retry(child)
             count += 1
     return count

--- a/loom-tools/tests/test_claude_config.py
+++ b/loom-tools/tests/test_claude_config.py
@@ -4,6 +4,7 @@ from __future__ import annotations
 
 import pathlib
 import shutil
+from unittest import mock
 
 import pytest
 
@@ -529,6 +530,48 @@ class TestCleanupAgentConfigDir:
     def test_returns_false_for_nonexistent(self, mock_repo: pathlib.Path) -> None:
         assert cleanup_agent_config_dir("nonexistent", mock_repo) is False
 
+    def test_tolerates_file_vanishing_during_rmtree(
+        self, mock_repo: pathlib.Path
+    ) -> None:
+        """Cleanup must not crash when a file (e.g. .claude.json.lock)
+        vanishes between rmtree listing the directory and unlinking the
+        entry.  This is the race condition from issue #3097.
+        """
+        import sys
+
+        setup_agent_config_dir("test-agent", mock_repo)
+        config_dir = mock_repo / ".loom" / "claude-config" / "test-agent"
+
+        # Create a fake lock file that Claude Code would leave behind
+        lock_file = config_dir / ".claude.json.lock"
+        lock_file.write_text("")
+
+        # Patch shutil.rmtree to simulate the race: the first call raises
+        # FileNotFoundError for the lock file (as if it vanished mid-removal),
+        # then the retry with ignore_errors succeeds.
+        original_rmtree = shutil.rmtree
+
+        def flaky_rmtree(path, *, onerror=None, onexc=None, ignore_errors=False, **kwargs):
+            if ignore_errors:
+                # Second (belt-and-suspenders) call — let it proceed normally
+                return original_rmtree(path, ignore_errors=True)
+            # First call — simulate a file vanishing mid-walk
+            exc = FileNotFoundError(
+                2, "No such file or directory", str(lock_file)
+            )
+            if sys.version_info >= (3, 12) and onexc is not None:
+                onexc(None, str(lock_file), exc)
+                return original_rmtree(path, ignore_errors=True)
+            elif onerror is not None:
+                onerror(None, str(lock_file), (FileNotFoundError, exc, None))
+                return original_rmtree(path, ignore_errors=True)
+            return original_rmtree(path)
+
+        with mock.patch("loom_tools.common.claude_config.shutil.rmtree", side_effect=flaky_rmtree):
+            result = cleanup_agent_config_dir("test-agent", mock_repo)
+
+        assert result is True
+
 
 class TestValidateAgentConfigDir:
     """Tests for validate_agent_config_dir.
@@ -738,3 +781,18 @@ class TestCleanupAllAgentConfigDirs:
     ) -> None:
         # Base dir doesn't exist yet
         assert cleanup_all_agent_config_dirs(mock_repo) == 0
+
+    def test_tolerates_file_vanishing_during_rmtree(
+        self, mock_repo: pathlib.Path
+    ) -> None:
+        """Same race condition as issue #3097 but for bulk cleanup."""
+        setup_agent_config_dir("agent-1", mock_repo)
+        config_dir = mock_repo / ".loom" / "claude-config" / "agent-1"
+        (config_dir / ".claude.json.lock").write_text("")
+
+        # The real _rmtree_with_retry handles the error via onerror callback,
+        # so we just verify that cleanup_all does not raise when lock files
+        # are present and can potentially vanish.
+        count = cleanup_all_agent_config_dirs(mock_repo)
+        assert count == 1
+        assert not config_dir.exists()

